lib/proxy/peer/quic/quic_test.go

```diff
@@ -0,0 +1,414 @@
+	})
+
+	serverTransport := newTransport(t)
+	server, err := NewServer(ServerConfig{
+		Dialer: noDialer,
+		GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+			return &serverCert, nil
+		},
+		GetClientCAs: func(*tls.ClientHelloInfo) (*x509.CertPool, error) {
+			return hostCAPool, nil
+		},
+	})
+	require.NoError(t, err)
+	go server.Serve(serverTransport)
+	t.Cleanup(func() { _ = server.Close() })
+
+	client, err := NewClientConn(ClientConnConfig{
+		Log:         slog.Default(),
+		PeerID:      serverHostID,
+		PeerAddr:    serverTransport.Conn.LocalAddr().String(),
+		ClusterName: "testcluster",
+		GetTLSCertificate: func() (*tls.Certificate, error) {
+			return &clientCert, nil
+		},
+		GetTLSRoots: func() (*x509.CertPool, error) {
+			return hostCAPool, nil
+		},
+		Transport: newTransport(t),
+	})
+	require.NoError(t, err)
+	t.Cleanup(func() { _ = client.Close() })
+
+	require.NoError(t, client.Ping(context.Background()))
+
+	conn, err := client.Dial(
+		uuid.NewString()+".testcluster",
+		&utils.NetAddr{
+			AddrNetwork: "tcp",
+			Addr:        "1.2.3.4:56",
+		}, &utils.NetAddr{
+			AddrNetwork: "tcp",
+			Addr:        "7.8.9.0:12",
+		},
+		types.NodeTunnel,
+	)
+	if err == nil {
+		_ = conn.Close()
+	}
+	require.ErrorAs(t, err, new(*trace.NotImplementedError))
+
+	server.dialer = peerDialerFunc(func(clusterName string, request peerdial.DialParams) (net.Conn, error) {
+		return nil, trace.NotFound("nope")
+	})
+	conn, err = client.Dial(
+		uuid.NewString()+".testcluster",
+		&utils.NetAddr{
+			AddrNetwork: "tcp",
+			Addr:        "1.2.3.4:56",
+		}, &utils.NetAddr{
+			AddrNetwork: "tcp",
+			Addr:        "7.8.9.0:12",
+		},
+		types.NodeTunnel,
+	)
+	if err == nil {
+		_ = conn.Close()
+	}
+	require.ErrorAs(t, err, new(*trace.NotFoundError))
+
+	pipeClose := make(chan struct{})
+	server.dialer = peerDialerFunc(func(clusterName string, request peerdial.DialParams) (net.Conn, error) {
+		if clusterName != "echo" {
+			return nil, trace.BadParameter("expected echo cluster")
+		}
+		if request.ServerID != "echo.echo" {
+			return nil, trace.NotFound("only echo.echo is here")
+		}
+		if request.ConnType != "echo" {
+			return nil, trace.CompareFailed("the only conntype is echo")
+		}
+		p1, p2 := net.Pipe()
+		go func() {
+			defer close(pipeClose)
+			defer p2.Close()
+			io.Copy(p2, p2)
+		}()
+		return p1, nil
+	})
+
+	conn, err = client.Dial(
+		"echo.echo",
+		&utils.NetAddr{
+			AddrNetwork: "tcp",
+			Addr:        "1.2.3.4:56",
+		}, &utils.NetAddr{
+			AddrNetwork: "tcp",
+			Addr:        "7.8.9.0:12",
+		},
+		"echo",
+	)
+	require.NoError(t, err)
+	t.Cleanup(func() { _ = conn.Close() })
+
+	n, err := conn.Write([]byte("abcd"))
+	require.NoError(t, err)
+	require.Equal(t, 4, n)
+
+	buf := make([]byte, 4)
+	n, err = io.ReadFull(conn, buf)
+	require.NoError(t, err)
+	require.Equal(t, 4, n)
+	require.Equal(t, "abcd", string(buf))
+
+	require.NoError(t, conn.Close())
+	t.Log("closed")
+	<-pipeClose
```

Comment:
Would it be a clean failure, or a timeout that doesn't respect the test timeout as specified by the harness?

Reply:
The global test timeout is usually too high, so I think it's worth adding an explicit 1s-2s here in case it locks for some reason. (As of now obviously this works, so it would have little effect.)